cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java

```diff
@@ -126,45 +155,121 @@
     this.streamCoordinatorClient = streamCoordinatorClient;
     this.privilegesManager = privilegesManager;
     this.authorizationEnforcer = authorizationEnforcer;
+    // TODO make maximum size and expire after write configurable?
+    this.appInjectors = CacheBuilder.newBuilder()
+      .maximumSize(10)
+      .expireAfterWrite(15, TimeUnit.MINUTES)
+      .removalListener(new RemovalListener<ApplicationId, Injector>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<ApplicationId, Injector> notification) {
+          Injector injector = notification.getValue();
+          if (injector != null) {
+            PreviewRuntimeService service = injector.getInstance(PreviewRuntimeService.class);
+            service.stopAndWait();
+          }
+          ApplicationId application = notification.getKey();
+          if (application == null) {
+            return;
+          }
+          java.nio.file.Path previewDirPath = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
```

Comment:
will this path always be local? even when we have preview in distributed mode in future?